### PR TITLE
API changes

### LINKED ISSUE #18: Match standard WebSocket API more closely
- [x] `ws$getState()` should be `ws$getReadyState()` or `ws$readyState()` or `ws$readyState`, and should reflect the numeric constants or "connecting", "open", "closing", "closed".
- [x] `onFail` should be `onError`
- [x] Event handlers should be assigned after construction (not passed into the constructor), and should be re-assignable
- [x] Constructor should have `protocols` argument which is used during handshake
- [x] Would prefer `WebsocketClient` R6 class be renamed `WebSocket` (note capitalization) to match standard JS API
- [x] `onClose`/`onError` handlers should be passed data about the close code/reason and error, respectively
- [x] Calling `ws$close()` before opening shouldn't throw an exception (but should raise onClose or onError)
- [x] Add `autoConnect = TRUE` constructor argument and `ws$connect()` method

### PR DESCRIPTION
**Pull request by jcheng5**
Fixes #18 ~except for `ws$close()` being available before connection has been established. I may or may not get to that for this release.~

~DO NOT MERGE until an accompanying PR is created for shinyloadtest.~

Now ready for merge, followed by https://github.com/rstudio/shinyloadtest/pull/29

## Testing notes

See the README.md once this is merged; the code example there should work.

### CODE REVIEW

**Comment by wch** [comment]
FWIW, we named it `WebsocketClient` instead of `WebSocket` because httpuv has a `WebSocket` class. https://github.com/rstudio/httpuv/blob/cb697dbffb6b2e1c650c6f23ea14e780bad62aa6/R/httpuv.R#L301

File: R/websocket.R [hunk 1]
```diff
@@ -108,33 +135,66 @@ null_func <- function(...) { }
...
+      autoConnect = TRUE,
       accessLogChannels = c("none"),
-      errorLogChannels = NULL
+      errorLogChannels = c("none")
     ) {
+      self$onOpen <-
+        self$onClose <-
+        self$onError <-
+        self$onMessage <-
+        function(event) {}
```
Comment (wch) [inline R/websocket.R]: This function will capture the environment of the `initialize` method, so it effectively isn't different from what was there before.

Comment (jcheng5) [reply R/websocket.R]: Yeah, that's fine. The references to these effectively get cleaned up during removeHandlers (in C++) so it doesn't really matter what they capture.

**Comment by jcheng5** [comment]
Shoot, why did we export httpuv::WebSocket??

**Comment by jcheng5** [comment]
Wondering now if the callbacks should work like ws$onOpen(callback) instead of ws$onOpen <- callback, in order to match httpuv better (and would also let us give a warning if you register onOpen/onMessage too late).

**Comment by jcheng5** [comment]
Resolved: `ws$onOpen(callback)` (etc.) and support multiple callback registrations.

**Comment by jcheng5** [comment]
@wch @alandipert This should be all set AFAICT.

**Comment by jcheng5** [comment]
@alandipert Phew, good catch! I could've sworn I had done that! Updated now.

**Pull request merged**